### PR TITLE
basic stack protection implementation

### PR DESCRIPTION
**Pull request by melentyev**
Implemented simple thread stack protection using mmu on x86.

### CODE REVIEW

File: src/kernel/thread/thread_allocator.c [hunk 1]
```diff
@@ -24,7 +25,7 @@ typedef union thread_pool_entry {
 	char stack[STACK_SZ];
 } thread_pool_entry_t;
 
-POOL_DEF(thread_pool, thread_pool_entry_t, POOL_SZ);
+POOL_DEF_ATTR(thread_pool, thread_pool_entry_t, POOL_SZ, __attribute__ ((aligned (4096))));
```
Comment (anton-bondarev) [inline src/kernel/thread/thread_allocator.c]: Use   VMEM_PAGE_SIZE instead of magic 4096

File: src/kernel/thread/stack_protect.c [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+}
+
+void stack_protect_disable() {
+    enabled = 0;
+}
+
+int stack_protect_enabled() 
+{
+    return enabled;
+}
```
Comment (anton-bondarev) [inline src/kernel/thread/stack_protect.c]: Please add a new line

File: src/include/kernel/thread/stack_protect.h [hunk 3]
```diff
@@ -0,0 +1,10 @@
+#include <mem/vmem.h>
+#include <kernel/thread.h>
+
+void stack_protect(struct thread *t, size_t size);
```
Comment (anton-bondarev) [inline src/include/kernel/thread/stack_protect.h]: Please use `extern` attribute for declarations

File: src/kernel/thread/stack_protect.c [hunk 4]
```diff
@@ -0,0 +1,49 @@
+#include <kernel/thread/stack_protect.h>
+
+#include <assert.h>
+#include <util/binalign.h>
+#include <hal/mmu.h>
+#include <mem/vmem/vmem_alloc.h>
+
+static int enabled = 0;
```
Comment (anton-bondarev) [inline src/kernel/thread/stack_protect.c]: Is this really required?
Do you want to protect from using your routines before mmu was enabled?

Comment (melentyev) [inline src/kernel/thread/stack_protect.c]: I just wanted to make protection configurable in runtime like vmem for example. Do you think it's unnecessary option?

File: src/kernel/thread/thread_allocator.c (hunk 1 above)
Comment (melentyev) [inline src/kernel/thread/thread_allocator.c]: Fixed

File: src/kernel/thread/Mybuild [hunk 5]
```diff
@@ -25,6 +25,8 @@ module core {
 
 	depends embox.kernel.cpu.common
 	depends embox.kernel.sched.current.api
+
+	depends stack_protect
```
Comment (anton-bondarev) [inline src/kernel/thread/Mybuild]: Please create `abstract module` with `@DefaultImpl` wich be empty implementation of your API.
Embox support a lot of platforms. Some of them haven't support MMU and if we'll accept your PR we'll break all of ones.

File: src/kernel/thread/thread_allocator.c [hunk 6]
```diff
@@ -24,7 +25,8 @@ typedef union thread_pool_entry {
 	char stack[STACK_SZ];
 } thread_pool_entry_t;
 
-POOL_DEF(thread_pool, thread_pool_entry_t, POOL_SZ);
+POOL_DEF_ATTR(thread_pool, thread_pool_entry_t, POOL_SZ, 
+    __attribute__ ((aligned (VMEM_PAGE_SIZE))));
```
Comment (anton-bondarev) [inline src/kernel/thread/thread_allocator.c]: Try to separate threads from VMEM modules (or add default implementation)

File: templates/x86/qemu/mods.config [hunk 7]
```diff
@@ -44,7 +44,7 @@ configuration conf {
 	include embox.kernel.thread.signal.siginfoq
 
 	@Runlevel(2) include embox.kernel.task.multi
-	@Runlevel(2) include embox.kernel.thread.core
+	@Runlevel(2) include embox.kernel.thread.core(thread_stack_size=16384)
```
Comment (anton-bondarev) [inline templates/x86/qemu/mods.config]: You shouldn't change a common template. It's used for many targets

File: src/kernel/thread/stack_protect.c (hunk 4 above)
Comment (anton-bondarev) [inline src/kernel/thread/stack_protect.c]: Ok, may be it enough for now

**Comment by anton-bondarev** [comment]
Please fix my latest comments and we'll accept your PR!

File: templates/x86/qemu/mods.config (hunk 7 above)
Comment (AntonKozlov) [inline templates/x86/qemu/mods.config]: For me, it's OK to have 16K user stack on x86.
Stack protection should pay off this overhead.

Comment (anton-bondarev) [inline templates/x86/qemu/mods.config]: Ok, if we want to use stack protection in this template by default I'm agree

File: src/kernel/thread/Mybuild (hunk 5 above)
Comment (melentyev) [inline src/kernel/thread/Mybuild]: Done. Is it ok?

**Comment by anton-bondarev** [comment]
Closed by #988

**Pull request closed without merging**